### PR TITLE
Rust: Implement enclosing callable

### PR DESCRIPTION
**Pull request by paldepind**
Fills out some more of the data flow implementation, including an implementation of `nodeGetEnclosingCallable`. This fixes a good deal of the `uniqueEnclosingCallable` inconsistencies. There is only 7 left post this PR.

### CODE REVIEW

File: rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll [hunk 1]
```diff
@@ -12,6 +12,64 @@ private import codeql.rust.controlflow.ControlFlowGraph
...
+}
+
+abstract class DataFlowCall extends TDataFlowCall {
+  /** Gets the enclosing callable. */
+  abstract DataFlowCallable getEnclosingCallable();
+
+  /** Gets the underlying source code call, if any. */
+  abstract CallCfgNode asCall();
+
+  abstract string toString();
```
Comment (hvitved) [inline rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll:52]: Add QL doc: `Gets a textual representation of this call.`

File: rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll [hunk 2]
```diff
@@ -115,11 +187,25 @@ module Node {
     override string toString() { result = node.toString() }
   }
 
-  final class ReturnNode extends NaNode {
-    RustDataFlow::ReturnKind getKind() { none() }
+  /** A data flow node that represents a value returned by a callable. */
+  final class ReturnNode extends ExprNode {
+    ReturnNode() { this.getCfgNode().getASuccessor() instanceof ExitCfgNode }
```
Comment (hvitved) [inline rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll:192]: I think this should just be `ReturnExpr`s.

File: rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll [hunk 3]
```diff
@@ -115,11 +187,25 @@ module Node {
...
+
+    ReturnKind getKind() { any() }
+  }
+
+  /** A data-flow node that represents the output of a call. */
+  abstract class OutNode extends Node, ExprNode {
+    /** Gets the underlying call, where this node is a corresponding output of kind `kind`. */
```
Comment (hvitved) [inline rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll:199]: Remove mention of `kind`.

File: rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll [hunk 4]
```diff
@@ -198,6 +284,12 @@ module LocalFlow {
   }
 }
 
+class DataFlowCallableAlias = DataFlowCallable;
+
+class ReturnKindAlias = ReturnKind;
+
+class DataFlowCallAlias = DataFlowCall;
```
Comment (hvitved) [inline rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll:291]: Make these aliases private.

File: rust/ql/test/extractor-tests/generated/Function/CONSISTENCY/DataFlowConsistency.expected [hunk 5]
```diff
@@ -1,2 +0,0 @@
-uniqueEnclosingCallable
```
Comment (hvitved) [inline rust/ql/test/extractor-tests/generated/Function/CONSISTENCY/DataFlowConsistency.expected:1]: This (and other) files can be removed now. I have used this script locally for removing empty consistency files:
```
find rust/ql/test -name "*Consistency.expected" -size 0 -print -delete
```

**Review by hvitved — CHANGES_REQUESTED** [review]
Great work!

File: rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll (hunk 2 above)
Comment (paldepind) [reply rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll:192]: Ok, so we don't want to include implicit returns like `fn foo() { 12 }`?

File: rust/ql/test/extractor-tests/generated/Function/CONSISTENCY/DataFlowConsistency.expected (hunk 5 above)
Comment (paldepind) [reply rust/ql/test/extractor-tests/generated/Function/CONSISTENCY/DataFlowConsistency.expected:1]: Nice, thanks 🙏

File: rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll (hunk 2 above)
Comment (hvitved) [reply rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll]: Right, we need to include that as well.

Comment (paldepind) [reply rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll:193]: I've reinstated the old implementation again :). If we want to include implicit returns then I think just looking at edges into the exit node handles all cases.
- `return` goes directly to the exit node
- In the absence of explicit returns, the `BlockExpr` is last both in a function and in a closure and it lets the final expression flow through if there is no `;` at the end.
- For a lambda without braces `|a| a + 1` the single expression has an edge to the exit node.

**Review by hvitved — APPROVED** [review]

**Pull request merged**